---
bool success
int32 num_elements
